cardano-chain-gen/src/Cardano/Mock/Forging/Tx/Generic.hs

```diff
@@ -0,0 +1,162 @@
+  ) where
+
+import           Cardano.Prelude hiding (length, (.))
+
+import           Data.List (nub)
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+import           Cardano.Ledger.Address
+import           Cardano.Ledger.BaseTypes
+import qualified Cardano.Ledger.Core as Core
+import           Cardano.Ledger.Credential
+import           Cardano.Ledger.Era (Crypto)
+import           Cardano.Ledger.Keys
+import           Cardano.Ledger.Shelley.LedgerState hiding (LedgerState)
+import           Cardano.Ledger.Shelley.TxBody
+import           Cardano.Ledger.Shelley.UTxO
+import           Cardano.Ledger.TxIn (TxIn (..))
+
+import           Ouroboros.Consensus.Cardano.Block (LedgerState)
+import           Ouroboros.Consensus.Shelley.Eras (StandardCrypto)
+import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock)
+import qualified Ouroboros.Consensus.Shelley.Ledger.Ledger as Consensus
+
+import           Cardano.Mock.Forging.Tx.Alonzo.ScriptsExamples
+import           Cardano.Mock.Forging.Types
+
+resolveAddress :: forall era. (Crypto era ~ StandardCrypto, HasField "address" (Core.TxOut era) (Addr (Crypto era)))
+               => UTxOIndex era -> LedgerState (ShelleyBlock era)
+               -> Either ForgingError (Addr (Crypto era))
+resolveAddress index st = case index of
+    UTxOAddressNew n -> Right $ Addr Testnet (unregisteredAddresses !! n) StakeRefNull
+    UTxOAddressNewWithStake n stakeIndex -> do
+      stakeCred <- resolveStakeCreds stakeIndex st
+      Right $ Addr Testnet (unregisteredAddresses !! n) (StakeRefBase stakeCred)
+    UTxOAddress addr -> Right addr
+    UTxOAddressNewWithPtr n ptr ->
+      Right $ Addr Testnet (unregisteredAddresses !! n) (StakeRefPtr ptr)
+    _ -> getField @"address" . snd . fst <$> resolveUTxOIndex index st
+
+resolveUTxOIndex :: forall era. (Crypto era ~ StandardCrypto, HasField "address" (Core.TxOut era) (Addr (Crypto era)))
+                 => UTxOIndex era -> LedgerState (ShelleyBlock era)
+                 -> Either ForgingError ((TxIn (Crypto era), Core.TxOut era), UTxOIndex era)
+resolveUTxOIndex index st = toLeft $ case index of
+    UTxOIndex n -> safeIndex n utxoPairs
+    UTxOAddress addr -> find (hasAddr addr) utxoPairs
+    UTxOInput input -> find (hasInput input) utxoPairs
+    UTxOPair pair -> Just pair
+    UTxOAddressNew _ -> do
+      addr <- rightToMaybe $ resolveAddress index st
+      find (hasAddr addr) utxoPairs
+    UTxOAddressNewWithStake _ _ -> do
+      addr <- rightToMaybe $ resolveAddress index st
+      find (hasAddr addr) utxoPairs
+    UTxOAddressNewWithPtr _ _ -> do
+      addr <- rightToMaybe $ resolveAddress index st
+      find (hasAddr addr) utxoPairs
+  where
+    utxoPairs :: [(TxIn (Crypto era), Core.TxOut era)]
+    utxoPairs = Map.toList $ unUTxO $ _utxo $ _utxoState $ esLState $
+        nesEs $ Consensus.shelleyLedgerState st
+
+    safeIndex n ls
+      | n < length ls = Just $ ls !! n
+      | otherwise = Nothing
```

Comment:
We already have `!?` from Data.List.Extra in our package deps, just use that one.